drivers/firmware/nrf_ironside/ironside_dvfs.c

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <hal/nrf_hsfll.h>
+#include <zephyr/kernel.h>
+
+#include <zephyr/drivers/firmware/nrf_ironside/ironside_dvfs.h>
+#include <zephyr/drivers/firmware/nrf_ironside/call.h>
+
+static volatile enum ironside_dvfs_oppoint current_dvfs_oppoint = IRONSIDE_DVFS_OPP_HIGH;
+static volatile enum ironside_dvfs_oppoint requested_dvfs_oppoint = IRONSIDE_DVFS_OPP_HIGH;
+K_MUTEX_DEFINE(ironside_dvfs_mutex);
+
+struct dvfs_hsfll_data {
+	uint32_t new_f_mult;
+	uint32_t new_f_trim_entry;
+	uint32_t max_hsfll_freq;
+};
+
+static const struct dvfs_hsfll_data dvfs_hsfll_data[IRONSIDE_DVFS_OPP_COUNT] = {
+	/* ABB oppoint 0.8V */
+	{
+		.new_f_mult = 20,
+		.new_f_trim_entry = 0,
+		.max_hsfll_freq = 320000000,
+	},
+	/* ABB oppoint 0.6V */
+	{
+		.new_f_mult = 8,
+		.new_f_trim_entry = 2,
+		.max_hsfll_freq = 128000000,
+	},
+	/* ABB oppoint 0.5V */
+	{
+		.new_f_mult = 4,
+		.new_f_trim_entry = 3,
+		.max_hsfll_freq = 64000000,
+	},
+};
+
+/**
+ * @brief Check if the requested oppoint is allowed.
+ *
+ * @param freq_setting The requested oppoint to check.
+ * @return true if the oppoint is allowed, false otherwise.
+ */
+static bool ironside_dvfs_opp_allowed(enum ironside_dvfs_oppoint freq_setting)
+{
+	if (freq_setting == IRONSIDE_DVFS_OPP_HIGH || freq_setting == IRONSIDE_DVFS_OPP_MEDLOW ||
+	    freq_setting == IRONSIDE_DVFS_OPP_LOW) {
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * @brief Check if the requested oppoint change operation is downscaling.
+ *
+ * @param target_freq_setting The target oppoint to check.
+ * @return true if the current oppoint is higher than the target, false otherwise.
+ */
+static bool ironside_dvfs_is_downscaling(enum ironside_dvfs_oppoint target_freq_setting)
+{
+	return current_dvfs_oppoint < target_freq_setting;
+}
+
+/**
+ * @brief Configure hsfll depending on selected oppoint
+ *
+ * @param enum oppoint target operation point
+ * @return 0 value indicates no error.
+ * @return -EINVAL invalid oppoint or domain.
+ */
+static int32_t ironside_dvfs_configure_hsfll(enum ironside_dvfs_oppoint oppoint)
+{
+	if (oppoint >= IRONSIDE_DVFS_OPP_COUNT) {
+		return -EINVAL;
+	}
+
+	nrf_hsfll_trim_t hsfll_trim = {};
+	uint8_t freq_trim_idx = dvfs_hsfll_data[oppoint].new_f_trim_entry;
+
+#if defined(NRF_APPLICATION)
+	hsfll_trim.vsup = NRF_FICR->TRIM.APPLICATION.HSFLL.TRIM.VSUP;
+	hsfll_trim.coarse = NRF_FICR->TRIM.APPLICATION.HSFLL.TRIM.COARSE[freq_trim_idx];
+	hsfll_trim.fine = NRF_FICR->TRIM.APPLICATION.HSFLL.TRIM.FINE[freq_trim_idx];
+#if NRF_HSFLL_HAS_TCOEF_TRIM
+	hsfll_trim.tcoef = NRF_FICR->TRIM.APPLICATION.HSFLL.TRIM.TCOEF;
+#endif
+#else
+	/* Currently only application core is supported */
+	return -EINVAL;
+#endif
```

Comment:
Couldn't this just be a `#error`? Better to fail in compile time than have this return an unexpected error at runtime.